Feature/tables and their CRUDs

S1mm1ssim opened the pull request.

S1mm1ssim committed:
Add dependencies

S1mm1ssim committed:
Setup postgres using docker-compose.yml

S1mm1ssim committed:
Implement migration scripts. Pass successfully

S1mm1ssim committed:
JDBC Datasource setup with HikariCP

S1mm1ssim committed:
Update schema: add foreign keys to position

S1mm1ssim committed:
Implement entities

S1mm1ssim committed:
Minor fix: rename field and add config param to constructor

S1mm1ssim committed:
Add migration: configure UUID to be generated in the DB

S1mm1ssim committed:
Implement controllers

S1mm1ssim committed:
Implement services

S1mm1ssim committed:
Implement DTOs and their corresponding mappers; add mapstruct dependency

S1mm1ssim committed:
Implement User's Role Enum

S1mm1ssim committed:
Implement EntityNotFoundException to handle 404 situations

S1mm1ssim committed:
Implement DAOs

S1mm1ssim committed:
Add postman collections

S1mm1ssim committed:
Add migrations

- Change user.role type to user_role enum
- Add unique index on user.email

S1mm1ssim committed:
Fix code style

S1mm1ssim committed:
Move SQL queries to constants

Versach commented on src/main/java/com/modsensoftware/marketplace/controller/CategoryController.java:
```diff
@@ -4,19 +4,26 @@
  * @author andrey.demyanchik on 11/3/2022
  */
-@AllArgsConstructor(onConstructor_ = {@Autowired})
+@RequiredArgsConstructor(onConstructor_ = {@Autowired})
```
onConstructor_ = {@Autowired} вроде и без этого работает нормально )

S1mm1ssim replied:
Да, работает и без нее. Я это пишу из соображений, что если Autowired всегда стоит добавлять на конструкторы просто, чтобы было более явно указано, что мы делаем injection, то и на ломбоковские аннотации тоже стоит добавлять.

Versach replied:
Кажется это все равно лишнее, инжекция через конструктор это общеизвестное поведение

Versach commented on src/main/java/com/modsensoftware/marketplace/controller/CategoryController.java:
```diff
@@ -33,7 +40,8 @@ public ResponseEntity<List<Category>> getAllCategories() {
 
     @GetMapping(value = "/{id}", produces = {"application/json"})
-    public ResponseEntity<Category> getCategoryById(@PathVariable(name = "id") Long id) throws EntityNotFoundException {
+    public ResponseEntity<Category> getCategoryById(@PathVariable(name = "id") Long id)
```
зачем тебе ResponseEntity?

S1mm1ssim replied:
В моем сетапе сейчас может и мало смысла в этом, но для отправки разных статусов мне кажется ResponseEntity самым удобным вариантом

Versach replied:
Для отправки разных статусов можно использовать ResponseCode, а если прям разные из одного метода ) кроме ошибок) - только в таком случае имеет смысл

Versach commented on src/main/java/com/modsensoftware/marketplace/dao/PositionDao.java:
```diff
@@ -20,11 +32,25 @@ public class PositionDao implements Dao<Position, Long> {
-                    "INNER JOIN item AS i ON p.item_id = i.id " +
-                    "INNER JOIN \"user\" AS u ON p.created_by = u.id " +
-                    "WHERE p.id = ?";
+            String query = "SELECT p.id AS position_id, p.item_id AS fk_position_item, "
```
повыноси эти запросы в константы, зачем ора внутри метода? итак дико тяжело читать

S1mm1ssim replied:
Я повыносил в последних коммитах, ты должно быть более ранние варианты видел.

Versach commented on src/main/java/com/modsensoftware/marketplace/dao/UserDao.java:
```diff
@@ -135,12 +153,13 @@ public void deleteById(UUID id) {
-                new Company(rs.getLong(9), rs.getString(10), rs.getString(11),
-                        rs.getTimestamp(12).toLocalDateTime(), rs.getString(13))
+        Company company = new Company(rs.getLong("company_id"), rs.getString("company_name"),
+                rs.getString("company_email"), rs.getTimestamp("company_created").toLocalDateTime(),
```
пожалуй тож в константы вынеси именя колонок, чекстайл не ругается ктстаи на повторения строк?

S1mm1ssim replied:
Хорошо, имена колонок вынесу. Нет, чекстайл не ругается. В паре мест идея только подсвечивает. Пожалуй это тоже вынесу в метод

S1mm1ssim replied:
Тут ты кстати тоже смотрел не финальный вариант кода. Я в последнем коммите причесал немного этот кусок

S1mm1ssim replied:
Дубликации кода избежать не сильно получается. На мой взгляд код становится ещё менее читаемым после этого. Например все методы deleteById между друг другом отличаются только SQL запросом, т.к. энтити разные. Есть небольшие куски по 7 строк, где в PreparedStatement сетаем значения на плейсхолдеры для UPDATE и INSERT. После того, как выносишь этот кусок, дубликация всё ещё остается. На мой взгляд будет проще оставить это дело и просто перейти на ORM. Это решит все проблемы, как с читаемостью и дубликацией. Но если ты настаиваешь на том, чтобы я это всё еще порефакторил, то я сделаю.

Versach commented on src/main/java/com/modsensoftware/marketplace/config/DataSource.java:
```diff
@@ -11,17 +11,17 @@
-        config.setUsername("postgres");
-        config.setPassword("postgres");
-        dataSource = new HikariDataSource(config);
+        CONFIG.setJdbcUrl("jdbc:postgresql://localhost:32768/marketplace");
```
Строго в проперти такие вещи вырносить

S1mm1ssim replied:
Этот кусок кода находится в static блоке. Я не нашел способа с помощью проперти все значения установить. Я пробовал через @Value на сеттерах делать, но приложение сразу ложится с IllegalArgumentException "dataSource or dataSourceClassName or jdbcUrl is required", потому что эти проперти null оказываются.

Versach commented on src/main/java/com/modsensoftware/marketplace/controller/CompanyController.java:
```diff
@@ -33,7 +40,8 @@ public ResponseEntity<List<Company>> getAllCompanies() {
     @GetMapping(value = "/{id}", produces = {"application/json"})
-    public ResponseEntity<Company> getCompanyById(@PathVariable(name = "id") Long id) throws EntityNotFoundException {
+    public ResponseEntity<Company> getCompanyById(@PathVariable(name = "id") Long id)
+            throws EntityNotFoundException {
```
А у тебя эксепшн checked чтоли?

S1mm1ssim replied:
Ой, случайно ошибся

Versach commented on src/main/java/com/modsensoftware/marketplace/dao/CompanyDao.java:
```diff
@@ -23,11 +23,17 @@
 @Component
 public class CompanyDao implements Dao<Company, Long> {
+
+    private static final String SELECT = "SELECT id, name, email, created, description FROM company";
```
nit: имя таблицы желательно тоже в отдельную константу выносить

S1mm1ssim committed:
Make EntityNotFoundException.java extend RuntimeException

S1mm1ssim committed:
Add maven-checkstyle plugin

S1mm1ssim committed:
Remove unused import

S1mm1ssim committed:
Make table names and their column names constant

S1mm1ssim committed:
Fix DataSource.java configuration

Use properties file to set up config

S1mm1ssim committed:
Make Position table name and its column names constant

S1mm1ssim committed:
Remove ResponseEntity from controller classes and use @ResponseStatus instead

S1mm1ssim committed:
Bugfix: setting CategoryDto parentId value to null did not have effect on already persisted category object

- Add nullParent property in CategoryDto to meet client's intentions on making parentId null
- Implement CategoryMapper without Mapstruct
- Fix null checks in CategoryDao.update()

S1mm1ssim committed:
Remove setIfNotNull in CategoryDao.update() when setting parentId due to NPE if parent was null previously

S1mm1ssim committed:
Merge pull request #2 from S1mm1ssim/add-checkstyle

Add checkstyle and refactor code

S1mm1ssim committed:
Merge branch 'develop' into feature/tables-crud
